packages/compiler/src/core/linter.ts

```diff
@@ -166,26 +185,62 @@ export function createLinter(
         rules: {},
       },
     };
+    const filteredRules = new Map<string, LinterRule<string, any>>();
+    for (const [ruleId, rule] of enabledRules) {
+      if ((rule.async ?? false) === asyncRules) {
+        filteredRules.set(ruleId, rule);
+      }
+    }
     tracer.trace(
       "lint",
-      `Running linter with following rules:\n` +
-        [...enabledRules.keys()].map((x) => ` - ${x}`).join("\n"),
+      `Running ${asyncRules ? "async" : "sync"} linter with following rules:\n` +
+        [...filteredRules.keys()].map((x) => ` - ${x}`).join("\n"),
     );
 
     const timer = startTimer();
-    for (const rule of enabledRules.values()) {
+    const exitCallbacks = [];
+    const allPromises: Promise<any>[] = [];
+    for (const rule of filteredRules.values()) {
       const createTiming = startTimer();
       const listener = rule.create(createLinterRuleContext(program, rule, diagnostics));
       stats.runtime.rules[rule.id] = createTiming.end();
       for (const [name, cb] of Object.entries(listener)) {
         const timedCb = (...args: any[]) => {
-          const duration = time(() => (cb as any)(...args));
-          stats.runtime.rules[rule.id] += duration;
+          const timer = startTimer();
+          const result = (cb as any)(...args);
+          if (isPromise(result)) {
```

Comment:
this check also should only matter for exit now

Reply:
changed to check event name and async flag, removed the isPromise util method. thanks.